crates/polars-time/src/truncate.rs

```diff
@@ -31,31 +25,61 @@ pub trait PolarsTruncate {
 impl PolarsTruncate for DatetimeChunked {
     fn truncate(
         &self,
-        options: &TruncateOptions,
         tz: Option<&Tz>,
+        every: &Utf8Chunked,
+        offset: &str,
         ambiguous: &Utf8Chunked,
     ) -> PolarsResult<Self> {
-        let every = Duration::parse(&options.every);
-        let offset = Duration::parse(&options.offset);
-        let w = Window::new(every, every, offset);
+        let offset = Duration::parse(offset);
 
         let func = match self.time_unit() {
             TimeUnit::Nanoseconds => Window::truncate_ns,
             TimeUnit::Microseconds => Window::truncate_us,
             TimeUnit::Milliseconds => Window::truncate_ms,
         };
 
-        let out = match ambiguous.len() {
-            1 => match ambiguous.get(0) {
-                Some(ambiguous) => self
-                    .0
-                    .try_apply(|timestamp| func(&w, timestamp, tz, ambiguous)),
-                _ => Ok(self.0.apply(|_| None)),
+        let out = match (every.len(), ambiguous.len()) {
+            (1, 1) => match (every.get(0), ambiguous.get(0)) {
+                (Some(every), Some(ambiguous)) => {
+                    let every = Duration::parse(&every);
+                    let w = Window::new(every, every, offset);
+                    self.0
+                        .try_apply(|timestamp| func(&w, timestamp, tz, ambiguous))
+                },
+                _ => Ok(Int64Chunked::full_null(self.name(), self.len())),
+            },
+            (1, _) => {
+                if let Some(every) = every.get(0) {
+                    let every = Duration::parse(&every);
+                    let w = Window::new(every, every, offset);
+                    try_binary_elementwise_values(
+                        self,
+                        ambiguous,
+                        |timestamp: i64, ambiguous: &str| func(&w, timestamp, tz, ambiguous),
+                    )
+                } else {
+                    Ok(Int64Chunked::full_null(self.name(), self.len()))
+                }
+            },
+            (_, 1) => {
+                if let Some(ambiguous) = ambiguous.get(0) {
+                    try_binary_elementwise(self, every, |opt_timestamp, opt_every| {
```

Comment:
I'm not totally sure I follow sorry - any chance you could show a test case where this makes a difference please?

Reply:
thanks for explaining - agree with your expected output

do you know why `None` is parsed as zero-duration? independently of this PR, perhaps that should be changed?

Reply:
`try_binary_elementwise` is better anyway. You should only use `try_binary_elementwise_values` if your function is very, very cheap (cheaper than a branch mispredict) and not worth skipping nulls for. Anything to do with parsing is way too expensive anyway and we should skip nulls for.

Reply:
I see, that's very clear - thanks both for explaining!

I understand that in the `(1, _)` arm, `try_binary_elementwise` isn't strictly necessary, but it would still be more performant, right?